keps/sig-node/4033-group-driver-detection-over-cri/README.md

```diff
@@ -0,0 +1,890 @@
+which cgroup driver to use. This removes the need for specifying cgroup driver
+in the kubelet configuration and eliminates the possibility of misaligned
+cgroup driver configuration between the kubelet and the runtime.
+
+## Motivation
+
+The responsibility of managing the Linux cgroups is currently split between the
+kubelet and the container runtime. Kubelet takes care of the pod (sandbox)
+level cgroups whereas the runtime is responsible for per-container cgroups.
+There currently are two different low-level management interfaces for cgroups:
+maninpulating the cgroupfs directly or using the systemd system management
+daemon to manage them. Currently, both the kubelet and the container runtime
+has a configuration setting for selecting the cgroup driver (cgroupfs or
+systemd). These settings must be in sync, both kubelet and the runtime
+configured to use the same driver as the two drivers are incompatible because
+of a different kind of cgroups hierarchy used in them. Having kubelet and the
+container runtime to use non-matching cgroup drivers can cause hard to
+understand failures in container creation or inconsistent resource allocation
+on the node. This – two independent configuration settings for the same thing –
+is a common cause for user errors. Instead of having a split brain situation,
+there should be a single source of truth for the cgroup driver.
+
+### Goals
+
+- make kubelet automatically use the same cgroup driver as the container
+  runtime, making it unnecessary to specify cgroup driver in kubelet
+  configuration
+- maintain backward compatibility to run kubelet on top of an older container
+  runtime that doesn’t support this new feature
+
+### Non-Goals
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a cluster administrator, I would like to simplify my node configuration by
+configuring fields just once, and not needing to synchronize between multiple
+processes.
+
+#### Story 2
+
+As a novice Kubernetes user, I would like to easily be able to start pods with
+all runtimes, even if they have differing opinions on defaults.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+Field adoption could be considered a risk, though the CRI implementations work
+closely with SIG Node and the feature will move along with CRI implementation
+adoption.
+
+## Design Details
+
+### CRI API
+
+Extend the CRI protocol to inform the kubelet which cgroup driver should be
+used.
+
+```diff
+ message RuntimeStatus {
+     // List of current observed runtime conditions.
+     repeated RuntimeCondition conditions = 1;
++    // Configuration settings of the runtime
++    RuntimeConfiguration configuration = 2;
```

Comment:
Regarding the runtime configuration we'd prolly need a "system-wide config" that is independent of the runtime class and the runtime handler specific stuff. I think the cgroup driver would fall in the first category – even though at least in containerd it is technically possible to have different cgroup driver for different runtime handlers I can't think of a practical scenario where this would make sense (and kubelet can anyways handle only one)

Also, I think we've tried to prefer lists with named fields over maps also in CRI API.

Thoughts @mrunalp @haircommander @mikebrow ?

Reply:
>  kubelet can anyways handle only one

👍 

Then this does not need to be a map, but I still expect the proto file to have a comment line to state that this is a global config that is decoupled from individual runtime handlers.

Reply:
I added a comment along these lines in the proposal (in the .proto diff)